openhands/controller/stuck.py

```diff
@@ -119,35 +124,49 @@ def _is_stuck_repeating_action_observation(self, last_actions, last_observations
+        # it takes 3 actions and 3 observations to detect a loop
+        # check if the last three actions are the same and result in errors
 
-        # are the last four actions the same?
-        if len(last_actions) == 4 and all(
-            self._eq_no_pid(last_actions[0], action) for action in last_actions
-        ):
-            # and the last four observations all errors?
-            if all(isinstance(obs, ErrorObservation) for obs in last_observations):
-                logger.warning('Action, ErrorObservation loop detected')
-                return True
-            # or, are the last four observations all IPythonRunCellObservation with SyntaxError?
-            elif all(
-                isinstance(obs, IPythonRunCellObservation) for obs in last_observations
-            ) and all(
-                cast(IPythonRunCellObservation, obs)
-                .content[-100:]
-                .find('SyntaxError: unterminated string literal (detected at line')
-                != -1
-                and len(
-                    cast(IPythonRunCellObservation, obs).content.split(
-                        'SyntaxError: unterminated string literal (detected at line'
-                    )[-1]
-                )
-                < 10
-                for obs in last_observations
+        if len(last_actions) >= 3 and len(last_observations) >= 3:
+            # are the last three actions the same?
+            if all(
+                self._eq_no_pid(last_actions[0], action) for action in last_actions[:3]
             ):
-                logger.warning('Action, IPythonRunCellObservation loop detected')
+                # and the last three observations all errors?
+                if all(
+                    isinstance(obs, ErrorObservation) for obs in last_observations[:3]
+                ):
+                    logger.warning('Action, ErrorObservation loop detected')
+                    return True
+                # or, are the last three observations all IPythonRunCellObservation with SyntaxError?
+                elif all(
+                    isinstance(obs, IPythonRunCellObservation)
+                    for obs in last_observations[:3]
+                ):
+                    for error_message in self.SYNTAX_ERROR_MESSAGES:
+                        if all(
+                            self._check_for_error_message(obs, error_message)
+                            for obs in last_observations[:3]
+                        ):
+                            logger.warning(
+                                'Action, IPythonRunCellObservation loop detected'
+                            )
+                            return True
+        return False
+
+    def _check_for_error_message(
+        self, obs: IPythonRunCellObservation, error_message: str
+    ) -> bool:
+        content = obs.content
+        # 4 lines, because the last 2 lines can be about interpreter and prompts
+        # since the introduction of the EventStreamRuntime
```

Comment:
Can be, or always are? Same for below `if error_message in line`, and above, `for error_message....check_for_error_message`:

The reason I ask is that this kind of hard-coded check here risks to trigger when the user is working with, lets say, files talking about syntax errors, files catching syntax errors, files like this file. I think the more strict, the more constraint we make it, the better we avoid the possibility of the stuck check triggering for an innocent case. So perhaps we can look if we can restrain the checks to be for exactly the line before last, or the one before it, whichever is the one with the error.
Not a must, dunno, maybe it's not possible; for the existing syntax error it was possible.

Reply:
Thanks for the feedback, happy to hear from you! 🙏 
Right now, it is "always are", for bash and ipython calls. Not sure, I can at that stage differentiate between them.
A more specific check for the last 2 lines would be tricky, as the prompts at runtime can vary and I don't want to duplicate a weird "sample" into this place, if that makes sense.